app/scripts/lib/stores/persistence-manager.ts

```diff
@@ -57,58 +114,186 @@ export class PersistenceManager {
+            const stringifiedBackup = JSON.stringify(backup);
+            // and the backup has changed
+            if (this.#backup !== stringifiedBackup) {
+              // save it to the backup DB
+              await this.#backupDb.set(backup);
+              this.#backup = stringifiedBackup;
+            }
+          }
+
+          if (this.#dataPersistenceFailing) {
+            this.#dataPersistenceFailing = false;
+          }
+        } catch (err) {
+          if (!this.#dataPersistenceFailing) {
+            this.#dataPersistenceFailing = true;
+            captureException(err);
+          }
+          log.error('error setting state in local store:', err);
+        } finally {
+          this.#isExtensionInitialized = true;
+        }
+      },
+    );
   }
 
-  async get() {
-    const result = await this.#localStore.get();
+  /**
+   * Retrieves the current state of the local store. If the store is empty,
+   * it returns undefined. If the store is not open, it throws an error.
+   *
+   * @returns The current state of the local store or null if the store is empty.
+   * @throws Error if the vault is missing and a backup vault is found in IndexedDB.
+   * @throws Error if the local store is not open.
+   */
+  async get(): Promise<MetaMaskStorageStructure | undefined> {
+    await this.open();
 
-    if (isEmpty(result)) {
-      this.#mostRecentRetrievedState = null;
-      return undefined;
+    return await navigator.locks.request(
+      STATE_LOCK,
+      { mode: 'shared' },
+      async () => {
+        const result = await this.#localStore.get();
+
+        // if we don't have a vault
+        if (!hasVault(result?.data)) {
+          // check if we have a backup in IndexedDB. we need to throw an error
+          // so that the user can be told about it. if we don't, carry on as if
+          // everything is fine (it might be, or maybe we lost BOTH the primary
+          // and backup vaults -- yikes!)
+          const backup = await this.getBackup();
+          // this check verifies if we have any keys saved in our backup.
+          // we use this as a sigil to determine if we've ever saved a vault
+          // before.
+          if (Object.values(backup).some((value) => value !== undefined)) {
+            // we've got some data (we haven't checked for a vault, as the
+            // background+UI are responsible for determining what happens now)
+            throw new Error(MISSING_VAULT_ERROR);
+          }
+        }
+
+        if (isEmpty(result)) {
+          this.#mostRecentRetrievedState = null;
+          return undefined;
+        }
+        if (!this.#isExtensionInitialized) {
+          this.#mostRecentRetrievedState = result;
+        }
+        return result;
+      },
+    );
+  }
+
+  async getBackup(): Promise<Backup> {
+    const [KeyringController, AppStateController, meta] =
+      await this.#backupDb.get([
+        'KeyringController',
+        'AppStateController',
```

Comment:
When we spoke, I had suggested we include `AppMetadataController` for diagnostic purposes. Was this what you intended to keep here, rather than `AppStateController`?

It's not clear to me why we'd want `AppStateController` unless we're trying to keep enough state to skip onboarding, and this alone isn't enough if we're attempting that.

Reply:
Oh, that is because

I'm a dummy. 😅